Introduce Cypress integration tests

gdemonet opened the pull request:
**Component**: ui, tests

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**: #2878 

**Summary**:

- Re-organize Cypress tests to isolate E2E tests
- Setup minimal mocks to run Cypress integration tests without a proper back-end
- Add complete test suite for the Navbar component
- Run Cypress integration tests in the CI

**Acceptance criteria**: 

- CI passes

---

<!-- Declare one or more issues to close once this PR gets merged -->

See: #2878

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

gdemonet committed:
ui/tests: Re-organize end-to-end Cypress tests

We move all end-to-end Gherkin-based tests to `cypress/integration/e2e`
(tried setting it up under `cypress/e2e`, but Cypress itself needs a
single location for spec files, so having everything under
`cypress/integration` makes it simpler).

We also rename feature files to be more generic, including multiple
scenarios in each when relevant (only for Solutions, at the moment),
and make sure to scope step definitions to their related features,
instead of putting everything under `common`.

We finally update the `test:e2e` and `test:e2esolutions` npm scripts to
use a more specific selection (only `.feature` files, and only under the
`e2e` directory).

See: #2878

gdemonet committed:
ui: Expose store and history to Cypress

This will allow easier manipulation of those objects (most notably for
stubbing methods) in Cypress tests.

See: #2878

gdemonet committed:
ui/cypress: Add minimal API mock fixtures

We define some useful fixtures for stubbing network responses that the
UI expects on load.
We also add a command, named `setupMocks`, to initialize this stubbing
with the aforementioned fixtures.

This serves as a first step, and will likely evolve according to the
various test needs.

See: #2878

gdemonet committed:
ui/cypress: Emulate an already logged in user

Instead of relying on an actual OIDC provider to manage the
authentication dance, we inject a JWT token in the browser's
LocalStorage to make the `oidc-client` believe we have an already logged
in user.
This logic is captured in a new command called `login`, which will make
it easy for various spec files to use it.

See: #2878

gdemonet committed:
ui/cypress: Add integration tests for navbar

We add a complete suite of tests to cover the various functionalities of
the navbar (various links, language selection, expected title, user
management) in isolation from other parts of the application.

Since this provides extended coverage of this component, we remove
redundant E2E tests used to validate language selection and link to
documentation.

See: #2878

ChengYanJin committed:
ui/ducks/layout: Fix the typo of sidebar_expanded

ChengYanJin committed:
ui/layout: Add data-cy for sidebar in order to do the interaction test

ChengYanJin committed:
ui/cypress: Add integration test for sidebar

gdemonet committed:
build/ui: Optimize caching of builder image

We were copying `src` and `public` while those are only needed at
runtime (when running `/entrypoint.sh build`). We also weren't running
`npm ci` at build-time, while we were already copying
`package[-lock].json`.

gdemonet committed:
ci/ui-unit-tests: Pre-install npm dependencies

We were running `npm ci` in a step, whereas this makes sense to be
cached in the worker image as much as possible (especially since we give
the full `ui/` folder as a build context).
We rely on `npm` cache to store as much as possible, and use it during
the CI run to make it faster (`--prefer-offline`).

See: #2878

gdemonet committed:
ci/ui: Add Cypress integration tests

We add these to the existing ui-unit-tests stage, thus installing
Cypress dependencies in the worker image.
Note that we need some complex management to handle starting the
application in the background and cleaning it up after Cypress tests are
complete (if we don't clean up, the Eve step hangs).

See: #2878

gdemonet committed:
ci/unit-tests: Mount workspace as emptyDir volume

This will improve I/O performance (compared to writing on the container
overlay fs).

bert-e commented:
# Hello gdemonet,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/2878-integration-tests` into
`development/2.6`
* [w/2.7/feature/2878-integration-tests](https://github.com/scality/metalk8s/compare/development/2.7...w/2.7/feature/2878-integration-tests) will be merged into `development/2.7` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality

bert-e commented:
# History mismatch



Merge commit #35d5e7a49ce091bec840a0256379d6e13bbfb6ed on the integration branch
`w/2.7/feature/2878-integration-tests` is merging a branch which is neither the current
branch `feature/2878-integration-tests` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `feature/2878-integration-tests` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet committed:
ci/ui-unit-tests: Increase worker resources

This will give more power for running Cypress (Electron-based) in a
reasonable amount of time.

See: #2878

gdemonet committed:
ci: Always copy artifacts to upload

This will silently fail on unknown artifacts, but should allow us to
retrieve logs when available.

ChengYanJin committed:
ui/tests: Add integration test for nodelist and nodetab

ChengYanJin committed:
ui/tests: Add baseURL in cypress.json

ChengYanJin committed:
ui/tests: Add the mock data in fixtures

gdemonet committed:
ui/cypress: Bump versions to latest

Bump Cypress to 5.5.0, cypress-cucumber-preprocessor to 4.0.0, and
cypress-wait-until to 1.7.1

See: #2878

gdemonet commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/2878-integration-tests` into
`development/2.6`
* [w/2.7/feature/2878-integration-tests](https://github.com/scality/metalk8s/compare/development/2.7...w/2.7/feature/2878-integration-tests) will be merged into `development/2.7` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality

ChengYanJin committed:
ui/cypress: Remove the env.target_url

Since we add baseUrl http://localhost:3000 in the cypress.json
we don't need the target_url anymore

ChengYanJin committed:
ui/tests: Add data-cy for the components for the tests

ChengYanJin committed:
ui/tests: Add general cy.route in cy.setupMocks()

ChengYanJin commented on ui/cypress/support/commands.js:
```diff
@@ -51,18 +51,18 @@ Cypress.Commands.add('setupMocks', () => {
 const ADMIN_JWT = {
   id_token:
     'eyJhbGciOiJSUzI1NiIsImtpZCI6IjY4MGZmMWVhZTBlZGYyMDI0MDQyOGVjMzRlMjBlODNjNzBlODAwODIifQ.eyJpc3MiOiJodHRwczovLzEwLjIwMC4yLjE1MDo4NDQzL29pZGMiLCJzdWIiOiJDaVF3T0dFNE5qZzBZaTFrWWpnNExUUmlOek10T1RCaE9TMHpZMlF4TmpZeFpqVTBOallTQld4dlkyRnMiLCJhdWQiOlsib2lkYy1hdXRoLWNsaWVudCIsIm1ldGFsazhzLXVpIl0sImV4cCI6MTYwMzQwMDg1MCwiaWF0IjoxNjAzMzE0NDUwLCJhenAiOiJtZXRhbGs4cy11aSIsIm5vbmNlIjoiZTM2MzRkMGQyYmRkNGMwMzg ...
-//   access_token:
```
Maybe we should uncomment these in the previous commits `ui/cypress: Emulate an already logged in user`?

ChengYanJin commented on ui/cypress/support/commands.js:
```diff
@@ -51,18 +51,18 @@ Cypress.Commands.add('setupMocks', () => {
-//     email: 'admin@metalk8s.invalid',
-//     email_verified: true,
-//     name: 'admin',
-//   },
```
Maybe we should uncomment these in the previous commits `ui/cypress: Emulate an already logged in user`?

gdemonet replied:
Oops you're right, bad fixup :)

ChengYanJin commented on ui/cypress/support/commands.js:
```diff
@@ -73,7 +73,7 @@ Cypress.Commands.add('login', () => {
         'oidc.user:/oidc:metalk8s-ui',
         JSON.stringify({
           ...ADMIN_JWT,
-        //   expires_at: Date.now() + 30000,
```
Same here?

ChengYanJin reviewed (commented):
Great works! Thank you! 

Only a small comment ^^

bert-e commented:
# History mismatch



Merge commit #334c79a09161826ea060cf88fd169eae251c56d5 on the integration branch
`w/2.7/feature/2878-integration-tests` is merging a branch which is neither the current
branch `feature/2878-integration-tests` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `feature/2878-integration-tests` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/2878-integration-tests` into
`development/2.6`
* [w/2.7/feature/2878-integration-tests](https://github.com/scality/metalk8s/compare/development/2.7...w/2.7/feature/2878-integration-tests) will be merged into `development/2.7` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality

bert-e commented:
# History mismatch



Merge commit #64d453ca6526e9e8c41e25aa3b323ce9ba6885f5 on the integration branch
`w/2.7/feature/2878-integration-tests` is merging a branch which is neither the current
branch `feature/2878-integration-tests` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `feature/2878-integration-tests` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/2878-integration-tests` into
`development/2.6`
* [w/2.7/feature/2878-integration-tests](https://github.com/scality/metalk8s/compare/development/2.7...w/2.7/feature/2878-integration-tests) will be merged into `development/2.7` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality

bert-e commented:
# History mismatch



Merge commit #cae053d1a81c6611ce575cf28889dc5abce2e74a on the integration branch
`w/2.7/feature/2878-integration-tests` is merging a branch which is neither the current
branch `feature/2878-integration-tests` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `feature/2878-integration-tests` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/2878-integration-tests` into
`development/2.6`
* [w/2.7/feature/2878-integration-tests](https://github.com/scality/metalk8s/compare/development/2.7...w/2.7/feature/2878-integration-tests) will be merged into `development/2.7` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality

bert-e commented:
# History mismatch



Merge commit #64a312be51a63c5f3243ea4bf32ea00be439d1ea on the integration branch
`w/2.7/feature/2878-integration-tests` is merging a branch which is neither the current
branch `feature/2878-integration-tests` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `feature/2878-integration-tests` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/2878-integration-tests` into
`development/2.6`
* [w/2.7/feature/2878-integration-tests](https://github.com/scality/metalk8s/compare/development/2.7...w/2.7/feature/2878-integration-tests) will be merged into `development/2.7` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality

gdemonet committed:
ci/build: Allocate even more RAM

We had issues with building the `metalk8s-ui` image in CI, likely due to
concurrent execution. The error was:

```
The build failed because the process exited too early. This probably
means the system ran out of memory or someone called `kill -9` on the
process.
```

To fix this, we add some memory to the Docker sidecar for the worker of
the `build` stage.

bert-e commented:
# History mismatch



Merge commit #b8f7fcf0896664add24a47c37878d940c4f928a6 on the integration branch
`w/2.7/feature/2878-integration-tests` is merging a branch which is neither the current
branch `feature/2878-integration-tests` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `feature/2878-integration-tests` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/2878-integration-tests` into
`development/2.6`
* [w/2.7/feature/2878-integration-tests](https://github.com/scality/metalk8s/compare/development/2.7...w/2.7/feature/2878-integration-tests) will be merged into `development/2.7` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality

ChengYanJin committed:
ui/tests: Create stubHistory command

ChengYanJin committed:
ui/test: Refactorize the code between source tabs

bert-e commented:
# History mismatch



Merge commit #db1d4762dfe45a91769da11db9b71d1104e2d06b on the integration branch
`w/2.7/feature/2878-integration-tests` is merging a branch which is neither the current
branch `feature/2878-integration-tests` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `feature/2878-integration-tests` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/2878-integration-tests` into
`development/2.6`
* [w/2.7/feature/2878-integration-tests](https://github.com/scality/metalk8s/compare/development/2.7...w/2.7/feature/2878-integration-tests) will be merged into `development/2.7` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality

ChengYanJin committed:
ui/tests: Improve the sidebar integration test

ChengYanJin reviewed (dismissed):
LGTM, thanks for the work!

gdemonet commented:
/approve

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/5c30d0e36e2f2eb1797d2259b9fef1c7f7eb529e) did not succeed in branch w/2.7/feature/2878-integration-tests.




*The following options are set:* **approve**

ChengYanJin committed:
ui/tests: Add window.gc()

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality










*The following options are set:* **approve**

bert-e commented:
# History mismatch



Merge commit #3d7fbb4d1463ba42c9fe04ebc23656aa8676dafe on the integration branch
`w/2.7/feature/2878-integration-tests` is merging a branch which is neither the current
branch `feature/2878-integration-tests` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `feature/2878-integration-tests` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

gdemonet commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/2878-integration-tests` into
`development/2.6`
* [w/2.7/feature/2878-integration-tests](https://github.com/scality/metalk8s/compare/development/2.7...w/2.7/feature/2878-integration-tests) will be merged into `development/2.7` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality










*The following options are set:* **approve**

ChengYanJin committed:
ui/tests: Add the viewportWidth and viewportHeight in cypress.json

In order to configure the browser to have the same viewport as real run application

ChengYanJin committed:
ui/tests: To make sure state is ready for the test

ChengYanJin committed:
ui/tests: Increase the default timeout

ChengYanJin committed:
ui/ducks/monitoring: Make sure the props form the query are ready before sending the request

During the tests, we discovered that there are a lot of query requests were sent
without waiting the props that form the queries ready.

Because of these, huge amount of requests were sent too soon slow down the test.

ChengYanJin committed:
ui/tests: Stub the history after click on the tab

Since we implement the sorting for table, we have `?sort=health` in the history

ChengYanJin committed:
ui/tests: Add more interaction test for nodelist

ChengYanJin committed:
ui/tests: Add/Rename data-cy in the component

ChengYanJin committed:
ui/tests: Add more interaction test for nodetab

ChengYanJin committed:
ui/tests: Add pvc and AM alerts in the fixtures

ChengYanJin committed:
ui/tests: Add cy.route to retrieve pv pvc in the command.js

ChengYanJin committed:
ui/tests: Add tests for volume list

ChengYanJin committed:
ui/tests: Add test for volumelist and volume tabs

gdemonet committed:
ui/cypress: Avoid calling cy.visit() twice

Calling `cy.visit` too soon after an initial page load will result in
some requests being sent and then being cancelled, leading to unexpected
behaviour in the application.

See: #2878

gdemonet committed:
ui/cypress: Ignore base URL for asserting location

We wanted to check the full URL seen by Cypress, but only checking the
`location.pathname` was sufficient. This will avoid issues (such as our
browser hiding the `:80` port section).

See: #2878

gdemonet committed:
ci/ui: Run application from a sidecar for Cypress

Instead of running the metalk8s-ui application using the development
server (which watches changes to the source files and hot-reloads), we
instead build an NGINX image containing the statically built application
and deploy it as a sidecar to the Eve worker.
To better isolate this effort, we add a new stage for running UI
integration tests.

See: #2878

gdemonet committed:
ui/cypress: Rework network stubbing

Using the experimental `cy.route2`, we build more generic route handlers
to cover all required requests made by the app during our interaction
tests.
We also remove some aliases and the corresponding `cy.wait` calls.

See: #2878

gdemonet committed:
ui/cypress: Introduce fetch polyfill

This will avoid using a strange hack (relies on our ability to stub
requests at the network layer, using `cy.route2`) to fallback on XHR.

See: #2878

gdemonet committed:
ui/cypress: Cancel requests after each test

We added an afterEach hook to call `window.gc`, though this doesn't fix
our issue with ongoing network requests being cancelled when new tests
start (which creates turmoil in the application).
We thus replace this hook with a more naive approach, which redirects
the browser to an `about:blank` page, to effectively cancel all ongoing
requests before a new test begins.

Note that this behaviour is preventing one from using the app after a
test case through Cypress runner.

See: #2878

bert-e commented:
# History mismatch



Merge commit #0c8536fd45551aad7f7eb822ee8d9e064ebcef60 on the integration branch
`w/2.7/feature/2878-integration-tests` is merging a branch which is neither the current
branch `feature/2878-integration-tests` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `feature/2878-integration-tests` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

gdemonet commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/2878-integration-tests` into
`development/2.6`
* [w/2.7/feature/2878-integration-tests](https://github.com/scality/metalk8s/compare/development/2.7...w/2.7/feature/2878-integration-tests) will be merged into `development/2.7` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality










*The following options are set:* **approve**

gdemonet committed:
ci/cypress: Use CYPRESS_BASE_URL to run tests

We were still using the `target_url` env variable, which was not read
anymore from the tests PoV.

See: #2878

gdemonet committed:
ui/tests: Bump Cypress to 5.6.0

This will allow to use recent experimental features for network
stubbing.
We take this opportunity to also update related dependencies.

See: #2878

gdemonet committed:
ui, ci: Remove unused cypress.sh script

This script isn't used by CI (which relies on the Terraform Bastion
being properly provisioned), and isn't useful for local testing (where
developers likely have everything installed already).

ChengYanJin committed:
ui/test: Adapt the test with the new UI changes

ChengYanJin committed:
ui/test: Make sure we don't call cy.visite() twice

ChengYanJin committed:
ui/test: Use cy.route2 to get alerts from AM

bert-e commented:
# History mismatch



Merge commit #7d96dab305e14872fd93eb22f695a0b7e45072d6 on the integration branch
`w/2.7/feature/2878-integration-tests` is merging a branch which is neither the current
branch `feature/2878-integration-tests` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `feature/2878-integration-tests` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

ChengYanJin committed:
Merge pull request #2929 from scality/feature/add-integration-test-for-volume-page

[UI/cypress] Add integration test for volume page

gdemonet commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/2878-integration-tests` into
`development/2.6`
* [w/2.7/feature/2878-integration-tests](https://github.com/scality/metalk8s/compare/development/2.7...w/2.7/feature/2878-integration-tests) will be merged into `development/2.7` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality










*The following options are set:* **approve**

gdemonet committed:
ci/ui: Wait for sidecar application to be ready

We had some flakies where the sidecar nginx wouldn't start soon enough.
This should avoid such issues in the future.

See: #2878

gdemonet commented:
/reset

bert-e commented:
# Lossy reset warning



There seems to be *manual* commits on integration branches (e.g. conflict
resolutions) that will be lost if you chose to *reset*.

**You can use the `force_reset` command if you still want me
to delete those branches.**





*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality










*The following options are set:* **approve**

gdemonet commented:
/force_reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/2878-integration-tests` into
`development/2.6`
* [w/2.7/feature/2878-integration-tests](https://github.com/scality/metalk8s/compare/development/2.7...w/2.7/feature/2878-integration-tests) will be merged into `development/2.7` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality










*The following options are set:* **approve**

gdemonet committed:
ui/cypress: Wait longer between tests

This may help with flaky issues around the Cypress embedded proxy server
(which is used for network stubbing): if a request to `/config.json`
goes through without being intercepted by this proxy server and returns
invalid values (e.g. `"api_url": "https://{ip}:8443/api/kubernetes"`),
the following requests will use an invalid URL which in turn crashes the
proxy server.

See: #2878

ChengYanJin reviewed (approved):
LGTM! 
Appreciate all the work!

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.6`

* :heavy_check_mark: `development/2.7`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.6`

* :heavy_check_mark: `development/2.7`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`



Please check the status of the associated issue None.

Goodbye gdemonet.
